store/store.go

```diff
@@ -597,88 +597,56 @@ type Search struct {
 	Query   string
 	Channel string
 	Private bool
+	Prefix  bool
 }
 
 // Find finds  (installable) snaps from the store, matching the
 // given Search.
 func (s *Store) Find(search *Search, user *auth.UserState) ([]*snap.Info, error) {
 	searchTerm := search.Query
-	channel := search.Channel
-	private := search.Private
 
-	if private && user == nil {
+	if search.Private && user == nil {
 		return nil, ErrUnauthenticated
 	}
 
-	// see https://github.com/snapcore/snapd/blob/master/docs/rest.md#v2find
-
 	searchTerm = strings.TrimSpace(searchTerm)
 
-	prefix := "name"
-	if private {
-		// default prefix to "text" for private, so e.g. `snap find --private foo` works.
-		prefix = "text"
-	}
-	exact := false
-	if idx := strings.IndexRune(searchTerm, ':'); idx >= 0 {
-		prefix = searchTerm[:idx]
-		searchTerm = searchTerm[idx+1:]
-
-		if prefix == "name" {
-			trimmed := strings.TrimSuffix(searchTerm, "*")
-			if len(trimmed) == len(searchTerm) {
-				exact = true
-			}
-			searchTerm = trimmed
-		}
-	}
-
 	if searchTerm == "" {
 		return nil, ErrEmptyQuery
 	}
 
-	if strings.ContainsAny(searchTerm, ":*") {
+	// these characters might have special meaning on the search
+	// server, and don't form part of a reasonable search, so
+	// abort if they're included.
+	//
+	// "-" might also be special on the server, but it's also a
+	// valid part of a package name, so we let it pass
+	if strings.ContainsAny(searchTerm, `+=&|><!(){}[]^"~*?:\/`) {
 		return nil, ErrBadQuery
 	}
 
-	// The store only supports "fuzzy" search for private snaps.
-	// See http://search.apps.ubuntu.com/docs/
-	if private && prefix != "text" {
-		return nil, ErrBadQuery
-	}
+	u := *s.searchURI // make a copy, so we can mutate it
+	q := u.Query()
 
-	if exact {
-		// this is actually a request for a single snap's details; short-cut out
-		snapInfo, err := s.Snap(searchTerm, channel, false, user)
-		switch err {
-		case nil:
-			return []*snap.Info{snapInfo}, nil
-		case ErrSnapNotFound:
-			// not finding something is an error for Snap but not for Find
-			return nil, nil
-		default:
-			return nil, err
+	if search.Private {
+		if search.Prefix {
+			// The store only supports "fuzzy" search for private snaps.
+			// See http://search.apps.ubuntu.com/docs/
+			return nil, ErrBadQuery
 		}
-	}
 
-	u := *s.searchURI // make a copy, so we can mutate it
-	q := u.Query()
+		q.Set("private", "true")
+	}
 
-	switch prefix {
-	case "name":
+	if search.Prefix {
 		q.Set("name", searchTerm)
-	case "text":
+	} else {
 		q.Set("q", searchTerm)
-		if private {
-			q.Set("private", "true")
-		}
-	default:
-		return nil, ErrBadPrefix
 	}
 
 	q.Set("confinement", "strict")
-	if channel != "" {
-		q.Set("channel", channel)
+	if search.Channel != "" {
+		q.Set("channel", search.Channel)
```

Comment:
apparently sending a channel to a general search is not even a thing anymore according to last conversation, not sure myself